File: src/main/java/io/outfox/typescriptpoet/PropertySpec.kt
/*
 * Copyright 2017 Outfox, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.outfox.typescriptpoet


class PropertySpec
private constructor(
   builder: Builder
) {
  val name = builder.name
  val type = builder.type
  val javaDoc = builder.javaDoc.build()
  val decorators = builder.decorators.toImmutableList()
  val modifiers = builder.modifiers.toImmutableSet()
  val initializer = builder.initializer
  val optional = builder.optional

  internal fun emit(
     codeWriter: CodeWriter,
     implicitModifiers: Set<Modifier>,
     asStatement: Boolean = false,
     withInitializer: Boolean = true
  ) {
    codeWriter.emitJavaDoc(javaDoc)
    codeWriter.emitDecorators(decorators, false)
    codeWriter.emitModifiers(modifiers, implicitModifiers)
    codeWriter.emitCode("%L${if (optional) "?" else ""}: %T", name, type)
    if (withInitializer && initializer != null) {
      codeWriter.emit(" = ")
      codeWriter.emitCode("%[%L%]", initializer)
    }
    if (asStatement) {
      codeWriter.emit(";\n")
    }
  }

  fun toBuilder(): Builder {
    val bldr = Builder(name, type, optional)
       .addJavadoc(javaDoc)
       .addDecorators(decorators)
       .addModifiers(*modifiers.toTypedArray())
    initializer?.let { bldr.initializer(it) }

    return bldr
  }

  class Builder internal constructor(
     internal val name: String,
     internal val type: TypeName,
     val optional: Boolean = false
  ) {
    internal val javaDoc = CodeBlock.builder()
    internal val decorators = mutableListOf<DecoratorSpec>()
    internal val modifiers = mutableListOf<Modifier>()
    internal var initializer: CodeBlock? = null

    fun addJavadoc(format: String, vararg args: Any) = apply {
      javaDoc.add(format, *args)
    }

    fun addJavadoc(block: CodeBlock) = apply {
      javaDoc.add(block)
    }


    fun addDecorators(decoratorSpecs: Iterable<DecoratorSpec>) = apply {
      decorators += decoratorSpecs
    }

    fun addDecorator(decoratorSpec: DecoratorSpec) = apply {
      decorators += decoratorSpec
    }

    fun addModifiers(vararg modifiers: Modifier) = apply {
      this.modifiers += modifiers
    }

    fun initializer(format: String, vararg args: Any?) = initializer(CodeBlock.of(format, *args))

    fun initializer(codeBlock: CodeBlock) = apply {
      check(this.initializer == null) { "initializer was already set" }
      this.initializer = codeBlock
    }

    fun build() = PropertySpec(this)
  }

  companion object {

    @JvmStatic
    fun builder(name: String, type: TypeName, optional: Boolean = false, vararg modifiers: Modifier): Builder {
      return Builder(name, type, optional).addModifiers(*modifiers)
    }

  }

}

File: src/PropertySpec.ts
import { imm, Imm } from 'ts-imm';
import { CodeBlock } from './CodeBlock';
import { CodeWriter } from './CodeWriter';
import { DecoratorSpec } from './DecoratorSpec';
import { Modifier } from './Modifier';
import { TypeName } from './TypeNames';

/** A generated property declaration. */
export class PropertySpec extends Imm<PropertySpec> {
  public static create(
    name: string,
    type: TypeName,
    optional: boolean = false,
    ...modifiers: Modifier[]
  ): PropertySpec {
    return new PropertySpec({
      name,
      type,
      javaDoc: CodeBlock.empty(),
      decorators: [],
      modifiers,
      initializerField: undefined,
      optional,
      implicitlyType: false,
    });
  }

  @imm
  public readonly name!: string;
  @imm
  public readonly type!: TypeName;
  @imm
  public readonly javaDoc!: CodeBlock;
  @imm
  public readonly decorators!: DecoratorSpec[];
  @imm
  public readonly modifiers!: Modifier[];
  @imm
  public readonly initializerField?: CodeBlock;
  @imm
  public readonly optional!: boolean;
  @imm
  public readonly implicitlyType!: boolean;

  public emit(
    codeWriter: CodeWriter,
    implicitModifiers: Modifier[] = [],
    asStatement: boolean = false,
    withInitializer: boolean = true
  ) {
    codeWriter.emitJavaDoc(this.javaDoc);
    codeWriter.emitDecorators(this.decorators, false);
    codeWriter.emitModifiers(this.modifiers, implicitModifiers);
    codeWriter.emitCode(this.name);
    if (this.optional) {
      codeWriter.emitCode('?');
    }
    if (!this.implicitlyType) {
      codeWriter.emitCode(`: %T`, this.type);
    }
    if (withInitializer && this.initializerField) {
      codeWriter.emit(' = ');
      codeWriter.emitCodeBlock(this.initializerField);
    }
    if (asStatement) {
      codeWriter.emit(';\n');
    }
  }

  public addJavadoc(format: string, ...args: any[]): this {
    return this.copy({
      javaDoc: this.javaDoc.add(format, ...args),
    });
  }

  public addJavadocBlock(block: CodeBlock): this {
    return this.copy({
      javaDoc: this.javaDoc.addCode(block),
    });
  }

  public addDecorators(...decoratorSpecs: DecoratorSpec[]): this {
    return this.copy({
      decorators: [...this.decorators, ...decoratorSpecs],
    });
  }

  public addDecorator(decoratorSpec: DecoratorSpec): this {
    return this.copy({
      decorators: [...this.decorators, decoratorSpec],
    });
  }

  public addModifiers(...modifiers: Modifier[]): this {
    return this.copy({
      modifiers: [...this.modifiers, ...modifiers],
    });
  }

  public setImplicitlyTyped(): this {
    return this.copy({ implicitlyType: true });
  }

  public initializer(format: string, ...args: any[]): this {
    return this.initializerBlock(CodeBlock.of(format, ...args));
  }

  public initializerBlock(codeBlock: CodeBlock): this {
    return this.copy({
      initializerField: codeBlock,
    });
  }

  public toString(): string {
    return CodeWriter.emitToString(this);
  }
}
